syntax = "proto3";

package example;

service ExampleService {
  rpc GetUser (GetUserRequest) returns (GetUserResponse);
  rpc CreateUser (CreateUserRequest) returns (CreateUserResponse);
  rpc UpdateUser (UpdateUserRequest) returns (UpdateUserResponse);
  rpc DeleteUser (DeleteUserRequest) returns (DeleteUserResponse);
  rpc ListUsers (ListUsersRequest) returns (ListUsersResponse);
}

message GetUserRequest {
  int32 user_id = 1;
}

message GetUserResponse {
  string username = 1;
  string email = 2;
}

message CreateUserRequest {
  string username = 1;
  string email = 2;
}

message CreateUserResponse {
  int32 user_id = 1;
}

message UpdateUserRequest {
  int32 user_id = 1;
  string new_email = 2;
}

message UpdateUserResponse {
  bool success = 1;
}

message DeleteUserRequest {
  int32 user_id = 1;
}

message DeleteUserResponse {
  bool success = 1;
}

message ListUsersRequest {}

message ListUsersResponse {
  repeated GetUserResponse users = 1;
}